Mathlib/CategoryTheory/Functor/KanExtension/Pointwise.lean

```diff
@@ -0,0 +1,191 @@
+we may say that `E` is a pointwise left Kan extension (`E.IsPointwiseLeftKanExtension`).
+
+Conversely, when `CostructuredArrow.proj L Y ⋙ F` has a colimit, we say that
+`F` has a pointwise left Kan extension at `Y : D` (`HasPointwiseLeftKanExtensionAt L F Y`),
+and if this holds for all `Y : D`, we construct a functor
+`pointwiseLeftKanExtension L F : D ⥤ H` and show it is a pointwise Kan extension.
+
+## TODO
+
+* obtain similar results for right Kan extensions
+* refactor the file `CategoryTheory.Limits.KanExtension` using this new general API
+
+## References
+* https://ncatlab.org/nlab/show/Kan+extension
+
+-/
+
+namespace CategoryTheory
+
+open Category Limits
+
+namespace Functor
+
+variable {C D H : Type*} [Category C] [Category D] [Category H] (L : C ⥤ D) (F : C ⥤ H)
+
+/-- The condition that a functor `F` has a pointwise left Kan extension along `L` at `Y`.
+It means that the functor `CostructuredArrow.proj L Y ⋙ F : CostructuredArrow L Y ⥤ H`
+has a colimit. -/
+abbrev HasPointwiseLeftKanExtensionAt (Y : D) :=
+  HasColimit (CostructuredArrow.proj L Y ⋙ F)
+
+/-- The condition that a functor `F` has a pointwise left Kan extension along `L`: it means
+that it has a pointwise left Kan extension at any object. -/
+abbrev HasPointwiseLeftKanExtension := ∀ (Y : D), HasPointwiseLeftKanExtensionAt L F Y
+
+namespace LeftExtension
+
+variable {F L} (E : LeftExtension L F)
+
+/-- The cocone for `CostructuredArrow.proj L Y ⋙ F` attached to `E : LeftExtension L F`.
+The point is this cocone is `E.right.obj Y` -/
+@[simps]
+def coconeAt (Y : D) : Cocone (CostructuredArrow.proj L Y ⋙ F) where
+  pt := E.right.obj Y
+  ι :=
+    { app := fun g => E.hom.app g.left ≫ E.right.map g.hom
+      naturality := fun g₁ g₂ φ => by
+        dsimp
+        rw [← CostructuredArrow.w φ]
+        simp only [assoc, NatTrans.naturality_assoc, Functor.comp_map,
+          Functor.map_comp, comp_id] }
+
+/-- A left extension `E : LeftExtension L F` is a pointwise left Kan extension at `Y` when
+`E.coconeAt Y` is a colimit cocone. -/
+def IsPointwiseLeftKanExtensionAt (Y : D) := IsColimit (E.coconeAt Y)
+
+variable {E}
+
+lemma IsPointwiseLeftKanExtensionAt.hasPointwiseLeftKanExtensionAt
+    {E : LeftExtension L F} {Y : D} (h : E.IsPointwiseLeftKanExtensionAt Y) :
+  HasPointwiseLeftKanExtensionAt L F Y := ⟨_, h⟩
+
+variable (E)
+
+/-- A left extension `E : LeftExtension L F` is a pointwise left Kan extension when
+it is a pointwise left Kan extension at any object. -/
+abbrev IsPointwiseLeftKanExtension := ∀ (Y : D), E.IsPointwiseLeftKanExtensionAt Y
+
+variable (h : E.IsPointwiseLeftKanExtension)
+
+lemma IsPointwiseLeftKanExtension.hasPointwiseLeftKanExtension :
+    HasPointwiseLeftKanExtension L F :=
+  fun Y => (h Y).hasPointwiseLeftKanExtensionAt
+
+/-- A pointwise left Kan extension is universal, i.e. it is a left Kan extension. -/
+def IsPointwiseLeftKanExtension.isUniversal : E.IsUniversal :=
+  IsInitial.ofUniqueHom (fun G => StructuredArrow.homMk
```

Comment:
Just a suggestion, feel free to ignore: you could consider factoring out the definition of the arrow and also of the cocone below in `pointwiseLeftKanExtension` in order to reduce nesting a bit.

Reply:
Thanks for the suggestion! I have introduced auxiliary definitions here, but not for `pointwiseLeftKanExtension`.